### ISSUE TITLE
False positive for stray end tags

### ISSUE THREAD

**Issue by LiLoDavis** [validator/validator#841]
Page: https://www.w3.org/WAI/demos/bad/before/home.html

The Nu HTML Checker reports two stray end tags on this page, but both have corresponding start tags.

Error: Stray end tag **noscript**.
From line 142, column 70; to line 142, column 80
`/FONT></B></noscript>↩    <`
_LL comment: The start tag is on the same line as the end tag (142)._

Error: Stray end tag **head**.
From line 144, column 3; to line 144, column 9
`script>↩  </head>↩  <bo`
_LL comment: The start tag is on line 1._

**Comment by LiLoDavis** [comment]
Perhaps related, I'm also getting false positives for unclosed elements on this page: https://www.uwb.edu/brand/website/accessibility/examples/inaccessible-page

**Comment by cvrebert** [comment]
For https://www.w3.org/WAI/demos/bad/before/home.html , I suspect it's related to the `<noscript>` containing illegal children due to it being within the `<head>`. Probably that causes the parser to implicitly close the `<noscript>` tag early, thus making the explicit close `</noscript>` tag extraneous.

https://html.spec.whatwg.org/multipage/scripting.html#the-noscript-element

> * In a `head` element, if scripting is disabled for the `noscript` element
>   * The `noscript` element must contain only `link`, `style`, and `meta` elements.
> * In a `head` element, if scripting is enabled for the `noscript` element
>   * The `noscript` element must contain only text, except that invoking the HTML fragment parsing algorithm with the `noscript` element as the context element and the text contents as the input must result in a list of nodes that consists only of `link`, `style`, and `meta` elements that would be conforming if they were children of the `noscript` element, and no parse errors.

`<b>` and `<font>` aren't among the permitted children.

**Comment by LiLoDavis** [comment]
Ok, so not exactly a false positive, but rather a false attribution.

**Comment by LiLoDavis** [comment]
Hm. If the problem is that a <noscript> in a <head> can't contain <b> or<font> elements, shouldn't removing the `<b>` and `<font>` elements from the `<noscript>` prevent the "stray end tag" error? It doesn't.

**Comment by sideshowbarker** [comment]
> Hm. If the problem is that a `<noscript>` in a `<head>` can't contain `<b>` or `<font>` elements, shouldn't removing the `<b>` and `<font>` elements from the `<noscript>` prevent the "stray end tag" error? It doesn't.

Can you doublecheck that?

Here is minimal test case:

```html
<!doctype html>
<HTML lang="">
<title>test</title>
<noscript><b></b></noscript>
```

That one has `<b>` in `<noscript>`, which makes the checker report *Stray end tag `noscript`*:

https://validator.w3.org/nu/?showsource=yes&doc=data%3Atext%2Fhtml%3Bcharset%3Dutf-8%2C%253C%2521doctype%2520html%253E%250D%250A%253CHTML%2520lang%253D%2522%2522%253E%250D%250A%253Ctitle%253Etest%253C%252Ftitle%253E%250D%250A%253Cnoscript%253E%253Cb%253E%253C%252Fb%253E%253C%252Fnoscript%253E#textarea

Here is another minimal test case:

```html
<!doctype html>
<HTML lang="">
<title>test</title>
<noscript></noscript>
```

That one has no `<b>` in `<noscript>`, and the checker reports no errors:

https://validator.w3.org/nu/?showsource=yes&doc=data%3Atext%2Fhtml%3Bcharset%3Dutf-8%2C%253C%2521doctype%2520html%253E%250D%250A%253CHTML%2520lang%253D%2522%2522%253E%250D%250A%253Ctitle%253Etest%253C%252Ftitle%253E%250D%250A%253Cnoscript%253E%253C%252Fnoscript%253E#textarea

**Comment by sideshowbarker** [comment]
> For https://www.w3.org/WAI/demos/bad/before/home.html , I suspect it's related to the `<noscript>` containing illegal children due to it being within the `<head>`. Probably that causes the parser to implicitly close the `<noscript>` tag early, thus making the explicit close `</noscript>` tag extraneous.

That is exactly the case. But it’s important to note that you’ll only see that behavior when scripting is disabled (as mentioned in the spec section you cited).

When scripting is *not* disabled — which of course in the normal case in web browsers — then the parser actually won’t implicitly close that `noscript` element. More specifically, when scripting is not disabled, the `<B><FONT COLOR=RED>This page uses scripts!!!</FONT></B>` inside the `NOSCRIPT` element in the source of https://www.w3.org/WAI/demos/bad/before/home.html just goes into the DOM as a text node.

But the checker is not capable of checking documents with scripting enabled. The checker doesn’t have a JavaScript engine to execute script with. So the checker uses the HTML parser in “scripting disabled” mode. And in “scripting disabled” mode, the parser doesn’t evaluate the `NOSCRIPT` content as text — instead the parser tries to parse any markup it finds inside the `NOSCRIPT`.

So exactly what happens here is that when the parser hits that `<b>` start tag, the parser inserts an implicit `</noscript>` end tag before the `<b>` start tag. But the parser doesn’t stop there; because the `b` element cannot appear in `head`, the parser also inserts both an implicit `</head>` end tag before the `<b>` start tag, and also inserts an implicit `<body>` start tag.

So with scripting disabled, this is what the parser ends up putting into the DOM:

<img width="588" alt="Screen Shot 2019-07-19 at 11 16 10" src="https://user-images.githubusercontent.com/194984/61504896-fa311080-aa17-11e9-9121-48d81b6ed903.png">

…and this is what ends up getting rendered:

<img width="512" alt="Screen Shot 2019-07-19 at 11 27 22" src="https://user-images.githubusercontent.com/194984/61504970-43816000-aa18-11e9-8310-d5eb597e0f36.png">

**Comment by sideshowbarker** [comment]
> Ok, so not exactly a false positive, but rather a false attribution.

Yeah, basically. The is just one of the many parts of the HTML parsing algorithm that is almost absurdly arcane and non-intuitive. So for this case, it’s very difficult to have the checker emit a user-friendly error message that clearly and succinctly explains what the root problem actually is.

**Comment by sideshowbarker** [comment]
> Perhaps related, I'm also getting false positives for unclosed elements on this page: https://www.uwb.edu/brand/website/accessibility/examples/inaccessible-page

I don’t get any messages about unclosed elements on that page —

https://validator.w3.org/nu/?doc=https%3A%2F%2Fwww.uwb.edu%2Fbrand%2Fwebsite%2Faccessibility%2Fexamples%2Finaccessible-page

Maybe something changed? It got updated in the meantime?

**Comment by LiLoDavis** [comment]
> > Perhaps related, I'm also getting false positives for unclosed elements on this page: https://www.uwb.edu/brand/website/accessibility/examples/inaccessible-page
> 
> I don’t get any messages about unclosed elements on that page —
> 
> https://validator.w3.org/nu/?doc=https%3A%2F%2Fwww.uwb.edu%2Fbrand%2Fwebsite%2Faccessibility%2Fexamples%2Finaccessible-page
> 
> Maybe something changed? It got updated in the meantime?

I've been using the **Check serialized DOM of current page** bookmarklet rather than typing the URL into the [Nu HTML Checker](https://validator.w3.org/nu/) page. I did it both ways just now and got different results. Wasn't expecting that. :-(
For that page:
- Using the Nu HTML Checker directly, I get 12 warnings and 2 errors. 
- Using the bookmarklet, I get 15 warnings, 14 errors, and 1 fatal error.

**Comment by sideshowbarker** [comment]
I think this is resolved per the comments above. If not, let me know and we can reopen it.
